proposals/kapp/001-crd-upgrade-safety/README.md

```diff
@@ -0,0 +1,107 @@
+This means preventing data loss and breaking of clients/workloads that may be using the API.
+Failure to safely update a CRD may result in:
+- Existing CustomResources (CRs) being invalidated, causing extensions to fail reconciliation of the existing CRs
+- Breaking clients expecting the old schema to be present. This may include other extensions, workloads, and tooling that depend on a specific version of the CRD.
+
+When an unsafe update occurs, it requires manual intervention to update all of the existing CRs to the new version of the CRD.
+
+## Terminology / Concepts
+- CustomResourceDefinition (CRD) - https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/
+- CustomResource (CR) - https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/
+- Fail Closed - In the scope of this document, "failing closed" is used to represent that if we can't explicitly validate something then we should fail.
+- Fail Open - In the scope of this document, "failing open" is used to represent that if we can't explicitly validate something then we should succeed.
+- Preflight Check - A preflight check is a validation that is run prior to performing some action. Typically, if the validation fails the action will not be performed.
+
+## Proposal
+In order to address the problem identified, it is proposed that a new, optional, preflight check is added to `kapp` to enable verifying that an `update` operation on a CRD is safe before attempting to apply any changes to the cluster.
+
+### Goals and Non-goals
+
+#### Goals
+- Providing `kapp` users a way to enable verification that an `update` operation on CRDs is safe and won't cause data loss or break clients
+
+#### Non-goals
+- Anything not pertaining to providing `kapp` users a way to ensure that `update` operations on CRDs are safe before applying changes
+
+### Specification
+A new preflight check will be added to `kapp` called `CRDUpgradeSafety`.
+
+#### `preflight` Package Changes
+The `CRDUpgradeSafety` preflight check is a very nuanced preflight check and different users will likely want to be able to configure this check a bit further to do things like:
+- Only issue warnings about potential problems
+- Fail closed/open
+
+In order to accomodate this configurability without introducing various versions of the same preflight check, the logic in https://github.com/carvel-dev/kapp/tree/develop/pkg/kapp/preflight will be updated to allow for configurable preflight checks using the [existing kapp configuration file](https://carvel.dev/kapp/docs/v0.60.x/config).
+
+As an example, configuring a preflight check could look something like:
+```yaml
+apiVersion: kapp.k14s.io/v1alpha1
+kind: Config
+...
+preflight:
+  - name: Check
+    config:
+      paramOne: somevalue
+      paramTwo: [1, 2, 3]
+  - name: OtherCheck
+    config:
+      otherParamOne:
+        value: foo
+...
+```
```

Comment:
The configuration file should be its own section of this proposal. It should also reference the `--preflight` option which overrides the enables. Something along the lines of...

#### Preflight Configuration

The `preflight` configuration is located in a kapp configuration file with a `preflight` section (or something). Each entry in `preflight` contains the `name` of the preflight check, and the optional `config` section is a map of configuration values for the preflight check. If a preflight check is specified, it is enabled; if a preflight check is not listed, it is disabled.

Configuring a non-existent preflight check is an error.

The command line flag `--preflight` contains a list of enabled preflight checks. This list overrides the enabled/disabled list of preflight checks in the configuration file. Preflight checks not listed in the `--preflight` command line flag are disabled, even if enabled in the configuration file.

All preflight checks must have reasonable default values, and must be able to work without configuration specified.

Reply:
Updated in https://github.com/carvel-dev/carvel/pull/729/commits/15cf9752853505ff265778af437d74b2e791ea25

Reply:
Marking this as resolved as it has been updated. Please feel free to leave any new review comments on the modifications - thanks!